base/pmap.jl

```diff
@@ -15,38 +19,166 @@ Note that `f` must be made available to all worker processes; see
+    if batch_size == 1
+        if distributed
+            f = remote(p, f)
+        end
+
+        if retry_n > 0
+            f = wrap_retry(f, retry_on_error, retry_n, retry_max_delay)
+        end
+        if on_error != nothing
+            f = wrap_on_error(f, on_error)
+        end
+        return (AsyncGenerator(f, c), nothing)
+    else
+        batches = batchsplit(c, min_batch_count = length(p) * 3,
+                                max_batch_size = batch_size)
+
+        # During batch processing, We need to ensure that if on_error is set, it is called
+        # for each element in error, and that we return as many elements as the original list.
+        # retry, if set, has to be called element wise and we will do a best-effort
+        # to ensure that we do not call mapped function on the same element more than retry_n.
+        # This guarantee is not possible in case of worker death / network errors, wherein
+        # we will retry the entire batch on a new worker.
+        f = wrap_on_error(f, (x,e)->BatchProcessingError(x,e); capture_data=true)
+        f = wrap_batch(f, p, on_error)
+        return (flatten(AsyncGenerator(f, batches)),
+                (p, f, results)->process_batch_errors!(p, f, results, on_error, retry_on_error, retry_n, retry_max_delay))
     end
-    batches = batchsplit(c, min_batch_count = length(p) * 3)
-    return flatten(AsyncGenerator(remote(p, b -> asyncmap(f, b)), batches))
 end
 
-pgenerate(p::WorkerPool, f, c1, c...) = pgenerate(p, a->f(a...), zip(c1, c...))
+pgenerate(p::WorkerPool, f, c1, c...; kwargs...) = pgenerate(p, a->f(a...), zip(c1, c...); kwargs...)
+
+pgenerate(f, c; kwargs...) = pgenerate(default_worker_pool(), f, c...; kwargs...)
+pgenerate(f, c1, c...; kwargs...) = pgenerate(a->f(a...), zip(c1, c...); kwargs...)
+
+function wrap_on_error(f, on_error; capture_data=false)
+    return x -> begin
+        try
+            f(x)
+        catch e
+            if capture_data
+                on_error(x, e)
+            else
+                on_error(e)
+            end
+        end
+    end
+end
 
-pgenerate(f, c) = pgenerate(default_worker_pool(), f, c...)
-pgenerate(f, c1, c...) = pgenerate(a->f(a...), zip(c1, c...))
+wrap_retry(f, on_error, n, max_delay) = retry(f, on_error; n=n, max_delay=max_delay)
+
+function wrap_batch(f, p, on_error)
+    f = asyncmap_batch(f)
+    return batch -> begin
+        try
+            remotecall_fetch(f, p, batch)
+        catch e
+            if on_error != nothing
+                return Any[BatchProcessingError(batch[i], e) for i in 1:length(batch)]
+            else
+                rethrow(e)
+            end
+        end
+    end
+end
+
+asyncmap_batch(f) = batch -> asyncmap(f, batch)
 
 
 """
-    pmap([::WorkerPool], f, c...) -> collection
+    pmap([::WorkerPool], f, c...; distributed=true, batch_size=1, on_error=nothing,
+                        retry_on_error=e->true, retry_n=0, retry_max_delay=0) -> collection
```

Comment:
I think genstdlib has issues when signatures in docstrings are wrapped, unfortunately